README.md

```diff
@@ -1,6 +1,826 @@
+    }
+```
+
+</details>
+
+---
+
+### 🥅 Goals
+
+- [ ] 🧪 Can you borrow and repay CORN?
+- [ ] ❓ What happens if you repay without having enough tokens to repay? Have you handled that well? (`Lending__RepayingFailed` might be nice to throw...)
+- [ ] 🛑 Can you borrow more than 120% of your collateral value? It should revert if you attempt this...
+
+---
+
+## Checkpoint 5: 📉 Liquidation Mechanism
+
+🏦 So we have a way to deposit collateral and borrow against it. Great! But what happens if the price of CORN goes up and the liquidation threshold is exceeded?
+
+🔨 We need a liquidation mechanism!
+
+🔍 Let's go to the `liquidate` function. We want anyone to be able to call this when a position is liquidatable. The caller must have enough CORN to repay the debt. This function should remove the borrower's debt AND the amount of collateral that is needed to cover the debt.
+
+⚠️ First let's make sure to revert if the user's position is not liquidatable with `Lending__NotLiquidatable`.
+
+🔄 Let's transfer the CORN to this contract from the liquidator and then burn it. (`transferFrom` and `burnFrom`).
+
+🧹 Clear the borrower's debt completely.
+
+🧮 Calculate the amount of collateral needed to cover the cost of the burned CORN and remove it from the borrower's collateral.
+> 💡 Keep in mind, It's not enough to simply have a liquidation mechanism. We need an incentive for people to trigger it!
+
+> ⚠️ We have simplified things by not adding any APY incentives (and inversely borrowing fees). These are important incentives in real lending markets that help to keep the market balanced by encouraging people to supply collateral or as a borrower to repay a loan that is requiring a high APR because the collateral is not as safe or nearing the liquidation threshold. These fees are a great place to add logic that generates protocol revenue as well by taking some of the borrowing APR and letting it accrue to the protocol's token, passing the rest along to the supplier. These incentives, along with the liquidation system, help to make sure there is always more value in collateral than value being borrowed.
+
+💰 **So** add the `LIQUIDATOR_REWARD` as a percentage on top of the collateral (but never exceeding the borrower's total collateral) so that the liquidator has a nice incentive to want to liquidate that poor borrower.
+
+💸 Transfer that amount of collateral to the liquidator.
+
+📢 Finally emit the `Liquidation` event.
+
+<details markdown='1'><summary>Solution Code</summary>
+
+```solidity
+    function liquidate(address user) public {
+        if (!isLiquidatable(user)) {
+            revert Lending__NotLiquidatable(); // Revert if position is not liquidatable
+        }
+
+        uint256 userDebt = s_userBorrowed[user]; // Get user's borrowed amount
+        uint256 userCollateral = s_userCollateral[user]; // Get user's collateral balance
+        uint256 collateralValue = calculateCollateralValue(user); // Calculate user's collateral value
+
```

Comment:
Above in the readme already mentioned that `The caller must have enough CORN to repay the debt`. But if he don't have and tries to liquidate someone from the frontend, it receives the unreadable error:

<img width="423" alt="Screenshot 2025-04-08 at 19 33 59" src="https://github.com/user-attachments/assets/8a23c23e-ae1a-4baa-8dfe-c79cb0e85dd0" />

Adding something like this here could help
```
if (i_corn.balanceOf(msg.sender) < userDebt) {
    revert Lending__InsufficientLiquidatorCorn();
}
```

so the error will be
<img width="409" alt="Screenshot 2025-04-08 at 19 43 53" src="https://github.com/user-attachments/assets/44335e0f-7136-4d45-bd13-441d6bd6a23d" />